plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,197 @@
+    }
+
+    /**
+     * Write a SearchQueryRecord to a StreamOutput
+     * @param out the StreamOutput to write
+     * @throws IOException IOException
+     */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeLong(timestamp);
+        out.writeMap(
+            measurements,
+            (stream, metricType) -> MetricType.writeTo(out, metricType),
+            (stream, measurement) -> Measurement.writeTo(out, measurement)
+        );
+        out.writeMap(attributes, (stream, attribute) -> Attribute.writeTo(out, attribute), StreamOutput::writeGenericValue);
+    }
+
+    /**
+     * Compare two SearchQueryRecord, based on the given MetricType
+     *
+     * @param a the first SearchQueryRecord to compare
+     * @param b the second SearchQueryRecord to compare
+     * @param metricType the MetricType to compare on
+     * @return 0 if the first SearchQueryRecord is numerically equal to the second SearchQueryRecord;
+     *        -1 if the first SearchQueryRecord is numerically less than the second SearchQueryRecord;
+     *         1 if the first SearchQueryRecord is numerically greater than the second SearchQueryRecord.
+     */
+    public static int compare(SearchQueryRecord a, SearchQueryRecord b, MetricType metricType) {
+        return a.getMeasurement(metricType).compareTo(b.getMeasurement(metricType));
+    }
+
+    /**
+     * Check if a SearchQueryRecord is deep equal to another record
+     * @param other the other SearchQueryRecord record
+     * @return true if two records are deep equal, false otherwise.
+     */
+    @SuppressWarnings("unchecked")
+    public boolean deepEquals(SearchQueryRecord other) {
+        if (this == other) {
+            return true;
+        }
+        Map<MetricType, Measurement<? extends Number>> measurements2 = other.getMeasurements();
+        Map<Attribute, Object> attributes2 = other.getAttributes();
+        if (timestamp != other.getTimestamp() || measurements.size() != measurements2.size() || attributes.size() != attributes2.size()) {
+            return false;
+        }
+        for (Map.Entry<MetricType, Measurement<? extends Number>> entry : measurements.entrySet()) {
+            MetricType metricType = entry.getKey();
+            Measurement<? extends Number> measurement = entry.getValue();
+            if (!measurements2.containsKey(metricType)
+                || !measurement.getName().equals(measurements2.get(metricType).getName())
+                || !measurement.getValue().equals(measurements2.get(metricType).getValue())) {
+                return false;
+            }
+        }
+        for (Map.Entry<Attribute, Object> entry : attributes.entrySet()) {
+            Attribute attribute = entry.getKey();
+            Object value = entry.getValue();
+            if (!attributes2.containsKey(attribute)) {
+                return false;
+            }
+            if (attribute.equals(Attribute.INDICES)) {
```

Comment:
i don't like these checks. Why do you need them? for unit tests? if yes, maybe move it there itself.

Reply:
Refactored this part to make the equal function concise.